[WIP] Exchange wflow

ncfrey opened the pull request:
## Summary
@utf 

The exchange wflow is an extension to the magnetism wflow that enables calculation of exchange parameters and critical temperatures in magnetic materials. 

* Map pre-computed magnetic orderings to a classical Heisenberg model and solve for exchange parameters to any desired order (nearest neighbor, next-nearest neighbor, etc.).
* Estimate an effective J parameter and mean field Tc from only the lowest energy FM and AFM orderings.
* Parse outputs and interface with VAMPIRE package to do Monte Carlo simulations of Tc.

## TODO (if any)

* The static calculation step should be noncollinear, with magnetic moments constrained along some axis and with magnitudes fixed at the values from the ground state ordering. 
* More computationally demanding and accurate methods for computing exchange should be included (e.g. RPA, magnetic force theorem, etc.)

utf commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+
+
+@explicit_serialize
+class HeisenbergModelMapping(FiretaskBase):
```
So this class seems to do two things at once. It computes the the exchange parameters AND adds the model to the database.

If the mapping is trivial then I think it would be best to just treat this is a ToDb FireTask (and rename it accordingly). If the mapping can take a considerable amount of time, say > 10 mins then maybe this could be broken up into two tasks. One to do the mapping and writing the mapping to a file, and other to add the data to the database.

ncfrey replied:
The timing for the mapping is highly variable so I'll split it into two tasks.

utf replied:
Ok great. Similar to the `BoltztrapToDb` firetask, you can just presume you are in the correct folder with the mapping written to a json file.

ncfrey replied:
Done! 👍

utf commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+    def run_task(self, fw_spec):
+
+        db_file = env_chk(self["db_file"], fw_spec)
+        wf_uuid = self["exchange_wf_uuid"]
```
In keeping with the magnetism workflow, can you rename this parameter in all of your firetasks to just wf_uuid...

utf commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+    """
+
+    required_params = ["db_file", "exchange_wf_uuid", "parent_structure", "avg"]
+    optional_params = []
```
You can remove this line.

utf commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+        exchange_wf_uuid (int): Unique id for record keeping.
+        parent_structure (Structure): Structure object with magmoms.
+        mc_settings (dict): A configuration dict for monte carlo.
+        avg (bool): Only <J> exchange param.
```
Rename this to `average`

utf commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+
+
+@explicit_serialize
+class VampireMC(FiretaskBase):
```
Again, if this is likely to take some time, can we break up the running and ToDb steps.

ncfrey replied:
Yes, I'll split them up.

utf commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+        mc_timesteps = mc_settings["mc_timesteps"]
+
+        # Get a converged Heisenberg model if one was found
+        # if fw_spec["converged_heisenberg_model"]:
```
Is there a reason why you don't use this approach?

ncfrey replied:
At the moment, because of the issues with the input static calculations, this sort of convergence testing is really not very informative or useful. I think with constrained magnetic moments it will be worth doing.

utf commented on atomate/vasp/fireworks/exchange.py:
```diff
@@ -0,0 +1,157 @@
+            parent_structure (Structure): Magnetic ground state.
+            parents (FireWorks): Parent FWs.
+            db_file (str): Path to file containing db credentials.
+            heisenberg_settings (dict): A config dict for Heisenberg model mapping.
```
Can you add more detail here with the supported options.

utf commented on atomate/vasp/fireworks/exchange.py:
```diff
@@ -0,0 +1,157 @@
+        db_file=DB_FILE,
+        mc_settings=None,
+        name="vampire caller",
+        c=None,
```
Again, remove.

utf commented on atomate/vasp/fireworks/exchange.py:
```diff
@@ -0,0 +1,157 @@
+        mc_settings=None,
+        name="vampire caller",
+        c=None,
+        avg=True,
```
Please rename to average throughout. We prefer explicit variable names where possible.

ncfrey replied:
Done!

utf commented on atomate/vasp/workflows/base/exchange.py:
```diff
@@ -0,0 +1,353 @@
+
+        Optionally, only the lowest energy FM and AFM configurations can be used to compute the average exchange interaction, <J>, without any static calculations.
+
+        Args:
```
Please add all args and kwargs to the docstring.

utf commented on atomate/vasp/workflows/base/exchange.py:
```diff
@@ -0,0 +1,353 @@
+            # Magnetic ordering analysis that generates 'magnetic_orderings'
+            # collection
+            fw_analysis = Firework(
+                ExchangeToDB(
```
What is the reason for using this versus MagneticOrderingsToDB?

utf commented on atomate/vasp/workflows/base/exchange.py:
```diff
@@ -0,0 +1,353 @@
+class ExchangeWF:
+    def __init__(
+        self,
+        magnetic_structures,
```
Can you add a check in the init that all structures have magnetism information and if not raise an error.

utf commented on atomate/vasp/workflows/base/exchange.py:
```diff
@@ -0,0 +1,353 @@
+__version__ = 0.1
+
+
+class ExchangeWF:
```
It seems like this overlaps considerably with `MagneticOrderingsWF`. Could this be simplified by adding an option to only perform static calculations in `MagneticOrderingsWF` and then use that internally to generate the most of the fireworks when `static=True`.

@mkhorton  I don't know if you have any thoughts on this?

ncfrey replied:
Yes, major overlap. The `MagneticOrderingsWF` could be modified to offer a static-only option, and also a "static-only with constrained noncollinear moments" option.

utf replied:
Ok. If it's not too much trouble to at least add the static only option I think that would help reduce the complexity of this pull request a lot.

utf replied:
Hi @ncfrey. I'm a bit confused about what's happened to this workflow, it seems like you remove the static option?

ncfrey replied:
Hi @utf, after discussing with @mkhorton it seemed like it would be easiest to kind of partition all magnetic ordering calculations into the `MagneticOrderingsWF`, so I moved the static only option there. 

We are going to discuss how that can be extended/modified to include the more robust options I've mentioned.

My idea was that now the `ExchangeWF` is totally agnostic to how the magnetic structures and energies are generated, which was a goal I had at the beginning. Let me know if anything is unclear.

utf replied:
That makes sense! Thanks for clearing that up.

ncfrey commented:
> Hi @ncfrey. Thanks for submitting this pull request. It looks in pretty good shape!
> 
> I've made some initial comments and suggestions. Please let me know if any of my comments aren't clear.

Awesome, thanks very much, Alex! I will get started on addressing these comments.

ncfrey commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+
+    Args:
+        db_file (str): path to the db file that holds your tasks
+        collection and that you want to hold the magnetic_orderings
```
Sorry about that, fixed now!

ncfrey commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+        "db_file",
+        "wf_uuid",
+        "parent_structure",
+        "perform_bader",
```
Added docstrings

ncfrey commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+
+        uuid = self["wf_uuid"]
+        db_file = env_chk(self.get("db_file"), fw_spec)
+        to_db = self.get("to_db", True)
```
Yes, fixed

ncfrey commented on atomate/vasp/firetasks/exchange_tasks.py:
```diff
@@ -0,0 +1,458 @@
+
+
+@explicit_serialize
+class ExchangeToDB(FiretaskBase):
```
Yes haha, I think the original plan was to add some additional info but that wasn't necessary and I never changed it. I'll get rid of it and switch to `MagneticOrderingsToDB`.

utf replied:
Ok! In that case, I think it would be useful to have the ability to distinguish magnetic orderings calculated with relaxations vs at a fixed geometry. Could you:

1. Implement the formatting and docstring changes I mentioned above to the `MagneticOrderingsToDB` firetask.
2. Add an option to `MagneticOrderingsToDb` called `additional_fields` which just gets added to the task doc. This is a common pattern in atomate, for example you can copy the implementation of `BoltztrapToDb`.
3. Make your workflow submit additional data indicating that the structure was not relaxed before hand. How you handle I don't have any opinions about.

That way we can easily filter out magnetic structures from the exchange workflow vs magnetism workflow.

@mkhorton Does that sound reasonable to you?

ncfrey replied:
I will go ahead and make these changes, it shouldn't be difficult and it will make it easier in the future to modify the static input step.

@utf and @mkhorton

mkhorton replied:
I'm on board with this, have been chatting with @ncfrey about our options offline, `MagneticOrderingsToDb` will need to be updated a little but otherwise it should be fine

ncfrey commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,352 @@
+        "db_file",
+        "wf_uuid",
+        "parent_structure",
+        "cutoff",
```
Yeah, I can grab those from the model instance 👍 I'll change that.

utf replied:
Ok great.  It also seems like the average parameter isn't required by the firetask any more.

ncfrey replied:
I cleaned that up but kept "cutoff" since it's used to distinguish between different .json models.

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,352 @@
+
+        db_file = env_chk(self["db_file"], fw_spec)
+        wf_uuid = self["wf_uuid"]
+        formula = self["parent_structure"].formula
```
It seems formula and formula_pretty are never used in this Firetask and parent structure is only required to get the formulas.

Can you remove these lines and also remove parent_structure as a required param.

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,352 @@
+    Args:
+        db_file (str): path to file containing the database credentials.
+        wf_uuid (int): Unique id for record keeping.
+        parent_structure (Structure): Structure object.
```
Can you double check whether parent_structure is actually required. Again, it seems like it is only used for formula_pretty which itself isn't used anywhere else.

If you do need the parent structure, can you get it from the database without requiring it as a parameter.

ncfrey replied:
Yep same thing, the model is being grabbed from the db and parent structure isn't needed. I'll take that out where it isn't needed.

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,352 @@
+    Args:
+        db_file (str): path to file containing the database credentials.
+        wf_uuid (int): Unique id for record keeping.
+        parent_structure (Structure): Structure object with magmoms.
```
Similar comment on parent_structure here.

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,352 @@
+
+        db_file = env_chk(self["db_file"], fw_spec)
+        wf_uuid = self["wf_uuid"]
+        formula_pretty = self["parent_structure"].composition.reduced_formula
```
And here... Can this not just be taken from the database.

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,336 @@
+    Args:
+        db_file (str): path to file containing the database credentials.
+        wf_uuid (int): Unique id for record keeping.
+        mc_settings (dict): A configuration dict for monte carlo.
```
Can you point to where the users can find the valid options. I.e., say something like. See pymatgen.x.y.VampireCaller for supported options.

utf replied:
Also can this be an optional argument? See comment below.

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,336 @@
+        #     hmodel = HeisenbergModel.from_dict(fw_spec["converged_heisenberg_model"])
+
+        vc = VampireCaller(
+            mc_box_size=mc_box_size,
```
Rather than expanding the mc_settings above, can you just do:

```python
vc = VampireCaller(
    hm=hmodel,
    avg=average,
    **mc_settings
)
```

That way if a user doesn't include one of the options, the VampireCaller will fall through back on the default settings.

Secondly, if you make `mc_settings` optional with a default value of an empty dict (`mc_settings = self.get("mc_settings, {})`)  this will again nicely support falling back on the default arguments.

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,336 @@
+
+        name = "heisenberg_model_" + str(self["cutoff"])
+        hmodel = loadfn(name + ".json")
+        #hmodel = HeisenbergModel.from_dict(hmodel_dict)
```
The loadfn should be the best way to do this. You can probably remove this commented line.

utf commented on atomate/vasp/fireworks/exchange.py:
```diff
@@ -0,0 +1,181 @@
+            parent_structure (Structure): Magnetic ground state.
+            parents (FireWorks): Parent FWs.
+            db_file (str): Path to file containing db credentials.
+            mc_settings (dict): A configuration dict for monte carlo.
```
I guess the type hinting for this should be `Optional[dict]`. And again, having this as option fits nicely in with the above!

utf commented on atomate/vasp/workflows/base/exchange.py:
```diff
@@ -0,0 +1,183 @@
+        )
+
+        # Vampire Monte Carlo
+        mc_settings = {
```
These are the same as the `VampireCaller` defaults. Currently this results in duplication of settings. Ideally if you change the default VampireCaller defaults to make them better, you'd then have to change them in atomate too. If you follow the above by making this an optional argument and then don't set it, we can use the Pymatgen defaults for free.

ncfrey replied:
Sounds good, I'll do the same with the `HeisenbergMapper` defaults as those are defined in pymatgen as well.

ncfrey commented:
> Hi @ncfrey
> 
> This is looking great and I think it's almost there.
> 
> What are your thoughts on adding a present workflow that automatically does the magnetism workflow and then feeds the results into your exchange workflow?

That sounds good to me. I think it might make sense to offer the option in the magnetism wflow to run exchange immediately as a follow on.

@mkhorton

utf commented:
> That sounds good to me. I think it might make sense to offer the option in the magnetism wflow to run exchange immediately as a follow on.

That makes sense as the exchange workflow stands. I'm worried once we start adding additional options to do constrained magnetic moments etc, the magnetism workflow might start getting a bit overloaded.

The other issue is that it somewhat goes against one of the principles of atomate, which is that the "base" workflows should be the bare minimum set of calculations needed to get a property and the "preset" workflows are fully featured workflows.

This distinction is a little fuzzy already with the magnetism workflow (there is no preset version), so I'm open to suggestions here.

ncfrey commented:
> > That sounds good to me. I think it might make sense to offer the option in the magnetism wflow to run exchange immediately as a follow on.
> 
> That makes sense as the exchange workflow stands. I'm worried once we start adding additional options to do constrained magnetic moments etc, the magnetism workflow might start getting a bit overloaded.
> 
> The other issue is that it somewhat goes against one of the principles of atomate, which is that the "base" workflows should be the bare minimum set of calculations needed to get a property and the "preset" workflows are fully featured workflows.
> 
> This distinction is a little fuzzy already with the magnetism workflow (there is no preset version), so I'm open to suggestions here.

Oh ok, yeah I think I see what you mean here. I'm going to talk to @mkhorton on Monday and we can discuss this. I think it would be a good idea in general to have some preset magnetism wflows anyway.

ncfrey commented:
Hey @utf, after discussing with @mkhorton what we're planning to do is keep this wflow and the magnetism wflow as they are, and then Guy Moore will build a preset wflow that combines them with the appropriate settings and the constrained calculations. How does that sound?

utf commented:
That sounds good to me. If you wouldn't mind implementing the last few comments I had about using the pymatgen defaults, I think this should be ready to merge

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,313 @@
+    """
+
+    required_params = [
+        "db_file",
```
It looks like `db_file` and `wf_uuid` are no longer required btw.

ncfrey replied:
Fixed!

ncfrey commented:
Hey @utf, not sure if you saw the latest commit - I took care of removing unnecessary parameters and defaulting to pymatgen defaults. Let me know if there are any more changes I should make.

utf commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,10 @@
+repos:
+-   repo: https://github.com/ambv/black
```
Can you update this: https://github.com/psf/black

utf replied:
Also, can you run all your new files through black once you're finished with the rest of the comments.

ncfrey replied:
No problem! I will get started on finishing up the wflow test.

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,309 @@
+        )
+
+        hmodels = [HeisenbergModel.from_dict(d["heisenberg_model"]) for d in docs]
+        hmodel = hmodels[0]  # Take the model with smallest NN cutoff
```
Is this guaranteed to be the model with the smallest cutoff? Should you sort the list on cutoff size first?

utf commented on atomate/vasp/workflows/tests/test_exchange_workflow.py:
```diff
@@ -0,0 +1,64 @@
+            db_file=self.db_file,
+        ).get_wf(c=c)
+
+        self.assertEqual(wf.name, "Mn3Al - Exchange")
```
Can you expand this into proper workflow testing. Currently the workflow is not run to check whether all fireworks/firetasks execute correctly. Once the workflow has finished you should examine the database at the end of the workflow to check that:

i) The workflow status is COMPLETED and not FIZZLED.
ii) The magnetism/exchange collections have the correct data in them.

See TestFerroelectricWorkflow for an example of how to do this (a lot of the complexity in that workflow is handling `RunFakeVasp` which you will not have in your workflow). Similar to your other tests, you can use `@requires` to avoid running the tests if vampire is not installed.

It would be good to get vampire installed on the testing environment. Is it available on conda?

ncfrey replied:
Unfortunately, I don't think it's available through conda. The build procedure is pretty simple though, and the C++ code is here: https://github.com/richard-evans/vampire

ncfrey commented:
@utf I added a real wflow test, following the format in TestFerroelectricWorkflow, that checks for correct data in the exchange collection and verifies that all the FWs are COMPLETED. I'm not sure if there's an easy way to add vampire to the testing environment, but let me know if there's anything I can do to help or any changes I should make to the tests.

utf commented:
Thanks for adding the test so quickly. It should be very easy to get vampire installed on circleCI. Can you add below this line https://github.com/hackingmaterials/atomate/blob/c00845908729740b5b55af9b54de94097b5899a6/.circleci/config.yml#L21

and below this line: 
https://github.com/hackingmaterials/atomate/blob/c00845908729740b5b55af9b54de94097b5899a6/.circleci/config.yml#L51

These lines:
```
wget https://vampire.york.ac.uk/resources/release-5/vampire-5.0-linux.tar.gz
tar -zxvf vampire-5.0-linux.tar.gz; cd linux; ./install.sh; cd ..
```

And then hopefully vampire tests will now run.

utf commented on .circleci/config.yml:
```diff
@@ -19,6 +19,9 @@ jobs:
             conda update --quiet --all
+            wget https://vampire.york.ac.uk/resources/release-5/vampire-5.0-linux.tar.gz
+            tar -zxvf vampire-5.0-linux.tar.gz
+            export PATH=/root/linux/:$PATH
```
Sorry about this, this line should be:

```
export PATH=/root/atomate/linux/:$PATH
```

utf commented on .circleci/config.yml:
```diff
@@ -49,6 +52,9 @@ jobs:
             conda update --quiet --all
+            wget https://vampire.york.ac.uk/resources/release-5/vampire-5.0-linux.tar.gz
+            tar -zxvf vampire-5.0-linux.tar.gz
+            export PATH=/root/linux/:$PATH
```
And so should this line:

```
export PATH=/root/atomate/linux/:$PATH
```

utf commented:
So the vampire tests now run!

They are flagging an error here:

```
self = <atomate.vasp.workflows.tests.test_exchange_workflow.TestExchangeWF testMethod=test_workflow>

    @unittest.skipIf(not vampire_present, "vampire not present")
    def test_workflow(self):
        wf = self._get_simulated_wflow()
    
        self.assertEqual(wf.name, "Mn3Al - Exchange")
    
        # Heisenberg + VampireMC = 2
        self.assertEqual(len(wf.fws), 2)
    
        fw_ids = self.lp.add_wf(wf)
        rapidfire(self.lp, fworker=FWorker(env={"db_file": self.db_file}))
    
        # Exchange collection
        # mmdb = VaspCalcDb.from_db_file(self.db_file, admin=True)
        # mmdb.collection = mmdb.db["exchange"]
        exchange_collection = self.get_task_database().exchange
    
        # Check Heisenberg model mapping
        d = exchange_collection.find_one({"task_name": "heisenberg model"})
>       self._check_run(d, "heisenberg model")

../../vasp/workflows/tests/test_exchange_workflow.py:78: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <atomate.vasp.workflows.tests.test_exchange_workflow.TestExchangeWF testMethod=test_workflow>
d = None, task_name = 'heisenberg model'

    def _check_run(self, d, task_name):
        if task_name == "heisenberg model":
>           self.assertEqual(d["nn_cutoff"], 3.0)
E           TypeError: 'NoneType' object is not subscriptable

../../vasp/workflows/tests/test_exchange_workflow.py:93: TypeError
```

ncfrey commented:
Good news that they are running! Hm, I'm not seeing that error when I run the tests locally. It looks like it's failing to grab the task doc from the `exchange` collection. Is there a reason why
`exchange_collection = self.get_task_database().exchange` would work locally but not in CI?

utf commented:
I don't think that is the issue.

If you check the test logs it has this:

```
020-04-08 16:08:43,032 INFO Performing db tune-up
2020-04-08 16:08:43,105 INFO LaunchPad was RESET.
2020-04-08 16:08:43,133 INFO Added a workflow. id_map: {-8408: 1, -8407: 2}
2020-04-08 16:08:43,141 INFO Created new dir /root/atomate/atomate/utils/scratch/launcher_2020-04-08-16-08-43-141564
2020-04-08 16:08:43,141 INFO Launching Rocket
2020-04-08 16:08:43,196 INFO RUNNING fw_id: 2 in directory: /root/atomate/atomate/utils/scratch/launcher_2020-04-08-16-08-43-141564
2020-04-08 16:08:43,224 INFO Task started: {{atomate.vasp.firetasks.exchange.HeisenbergModelMapping}}.
2020-04-08 16:08:43,687 INFO Task completed: {{atomate.vasp.firetasks.exchange.HeisenbergModelMapping}} 
2020-04-08 16:08:43,767 INFO Rocket finished
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/fireworks/core/rocket.py", line 245, in run
    Rocket.update_checkpoint(lp, launch_dir, launch_id, checkpoint)
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/fireworks/core/rocket.py", line 435, in update_checkpoint
    launchpad.ping_launch(launch_id, checkpoint=checkpoint)
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/fireworks/core/launchpad.py", line 1525, in ping_launch
    m_launch.trackers]}})
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/pymongo/collection.py", line 1003, in update_one
    session=session),
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/pymongo/collection.py", line 856, in _update_retryable
    _update, session)
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/pymongo/mongo_client.py", line 1491, in _retryable_write
    return self._retry_with_session(retryable, func, s, None)
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/pymongo/mongo_client.py", line 1384, in _retry_with_session
    return func(session, sock_info, retryable)
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/pymongo/collection.py", line 852, in _update
    retryable_write=retryable_write)
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/pymongo/collection.py", line 823, in _update
    _check_write_command_response(result)
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/pymongo/helpers.py", line 221, in _check_write_command_response
    _raise_last_write_error(write_errors)
  File "/root/miniconda3/envs/test_env/lib/python3.7/site-packages/pymongo/helpers.py", line 203, in _raise_last_write_error
    raise WriteError(error.get("errmsg"), error.get("code"), error)
pymongo.errors.WriteError: The dotted field 'heisenberg_model_3.0' in 'state_history.0.checkpoint._all_update_spec.heisenberg_model_3.0' is not valid for storage.
```

So it seems like the string "heisenberg_model_3.0" is somewhere in firetask and is causing confusion.

utf commented on atomate/vasp/firetasks/exchange.py:
```diff
@@ -0,0 +1,289 @@
+        hmodel_dict = hmodel.as_dict()
+
+        # Update FW spec with model
+        name = "heisenberg_model_" + str(hmodel.cutoff)
```
This is the line that is causing confusion.

Can you edit this line to be
```python
name = "heisenberg_model_" + str(hmodel.cutoff).replace(".", "_")
```

ncfrey replied:
Got it, thanks!

utf commented:
Thanks for all your work on this pull request.  Everything looks good now, so I'll merge.
